Java/Android : Insecure Context Creation

porcupineyhairs opened the pull request:
Using untrusted input to create a context can lead to arbitary code execution
This PR adds a query to detect cases where an untrusted input may be
used to instantiate a new context in the application's scope. Allowing
this results in arbitrary code execution as a malicious app would now be
able to take control of the execution flow.

smowton commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qll:
Use `FieldAccess`, and check the name and type of the field being accessed. Example: https://github.com/github/codeql/blob/main/java/ql/src/Likely%20Bugs/Arithmetic/BadCheckOdd.ql#L22

smowton reviewed (commented):
Let's have some tests for this

porcupineyhairs commented:
@smowton I have included the tests and changes from the review. 

Please note, while I have somehow managed to get the tests working this time. Not having an automated stubber makes it difficult to stub android dependencies manually. Hence, I won't be including tests with my future Android PR's for now.

smowton commented on java/ql/test/query-tests/security/CWE-691/semmle/tests/InsecureClassLoading.java:
```diff
@@ -0,0 +1,34 @@
+  private void invokePlugins() {
+    PackageManager pm;
+    for (PackageInfo info : pm.getInstalledPackages(0)) {
+      // for (PackageInfo info : getPackageManager().getInstalledPackages(0)) {
```
?

porcupineyhairs replied:
Removed.

smowton reviewed (commented):
Mostly looks good, will tag a codeowner for review

aschackmull reviewed (changes requested):
Please add query help (did you forget to commit the `.qhelp` file?) and move the test to the `experimental` folder. Also, what's the deal with all the deletions in `Context.java`?

porcupineyhairs commented:
@aschackmull I have moved the tests and added the qhelp. As for the changes in `Context.java` they are intentional. I have removed the comments and lot of missing definitions.

smowton commented:
@aschackmull I've made an LGTM evaluation run for this and posted it on the bounty ticket

porcupineyhairs commented:
@smowton @aschackmull Can you share the LGTM run results with me on slack too?

aschackmull commented on java/ql/src/semmle/code/java/dataflow/FlowSources.qll:
```diff
@@ -341,3 +341,16 @@ class ExportedAndroidIntentInput extends RemoteFlowSource, AndroidIntentInput {
+
+/** Android `PackageInfo` objects that may have come from a hostile application. */
+class AndroidPackageInput extends DataFlow::Node {
+  Type receiverType;
+
+  AndroidPackageInput() {
+    exists(MethodAccess ma |
+      ma.getMethod().getQualifiedName() = "PackageManager.getInstalledPackages"
+    |
+      this.asExpr() = ma
+    )
```
Please use `RefType.hasQualifiedName` instead of `Method.getQualifiedName` to match the type, as the latter doesn't include the package name. Also, the field is unused.
```suggestion
  AndroidPackageInput() {
    exists(MethodAccess ma, Method m |
      ma.getMethod() = m and
      m.getDeclaringType().hasQualifiedName("android.content.pm", "PackageManager") and
      m.hasName("getInstalledPackages") and
      this.asExpr() = ma
    )
```

aschackmull commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qll:
```diff
@@ -0,0 +1,49 @@
+
+/** Holds if a taint flows from `node1` to `node2` through an access on the `PackageInfo` object */
+predicate packageInfoAccessStep(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(FieldAccess fa | fa.getQualifier() = node1.asExpr() and fa.getField() instanceof PackageNameField | fa =node2.asExpr()
```
Needs autoformat.

aschackmull commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qll:
```diff
@@ -0,0 +1,49 @@
+/** A sink for insecure context creation vulnerabilities */
+class InsecureContextCreationSink extends DataFlow::ExprNode {
+  InsecureContextCreationSink() {
+    exists(MethodAccess ma |
+      ma.getMethod().getQualifiedName().matches("ContextWrapper.create%Context")
+    |
```
```suggestion
    exists(MethodAccess ma, Method m |
      ma.getMethod() = m and
      m.getDeclaringType().hasQualifiedName("android.content", "ContextWrapper") and
      m.getName().matches("create%Context") and
```

aschackmull commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qll:
```diff
@@ -0,0 +1,49 @@
+/**
+ * A taint-tracking configuration for insecure context creation vulnerabilities
+ */
+class InsecureContextCreationConfig extends TaintTracking::Configuration {
```
Configurations should in general be defined only in query files (`.ql`), so this should be moved into `InsecureClassLoading.ql`.

aschackmull replied:
If you want, you can simply move all of this file into the query file.

aschackmull commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qll:
```diff
@@ -0,0 +1,49 @@
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking2
```
Might as well make this `TaintTracking`.
```suggestion
import semmle.code.java.dataflow.TaintTracking
```

aschackmull commented on java/ql/test/query-tests/maven-dependencies/.settings/org.eclipse.m2e.core.prefs:
```diff
@@ -0,0 +1,4 @@
+activeProfiles=
```
This file looks like it has been committed by accident. Please remove.

aibaars commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qhelp:
```diff
@@ -0,0 +1,33 @@
+      <li> 
+        To avoid arbitrary code execution, the application should sanitize all external inputs.
+      </li>
+      <li> Changing the application architecture to prevent unintended access to sensitive components.</li>
```
```suggestion
      <li> Changing the application architecture to prevent unintended access to sensitive components.</li>
    </ol>
```

aibaars commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qhelp:
```diff
@@ -0,0 +1,33 @@
+  <example>
+    <p>The following example shows a 'bad' way of creating Android `Context` variables. An attacker may register a package with the name `somename.plugin.` having a `pluginLoader`  class.</p>
+    <sample src="InsecureClassLoading.java" />
+    <p> Then, when the code above executes, the attacker would be able to run malicious code in the context of the applciaiton.</p>
```
```suggestion
    <p> Then, when the code above executes, the attacker would be able to run malicious code in the context of the application.</p>
```

aibaars replied:
@porcupineyhairs : don't forget about this typo.

smowton replied:
^^

aibaars commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qhelp:
```diff
@@ -0,0 +1,33 @@
+    <p> Then, when the code above executes, the attacker would be able to run malicious code in the context of the applciaiton.</p>
+  </example>
+
+  <references>
+    <li>
+      <a href="https://cwe.mitre.org/data/definitions/749.html">CWE-749</a>
+    </li>    
+  </references>
```
If I recall correctly CWE links are  extracted from the `@tags` property on the query and automatically included when rendering a QHelp file . In addition you link to CWE-749 in this file but the query `@tags` mentions CWE-691 instead

```suggestion
```

aschackmull replied:
For that to work I think there needs to be a reference section - possibly with at least one other reference in it (which is good practice anyway).

aibaars replied:
Ah curious, didn't know that. It seems to work fine with an empty references section. 

@porcupineyhairs You can "render" a qhelp file using the following command:
```
codeql generate query-help --format=markdown src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qhelp
```

aibaars replied:
Below is a rendered version of the QHelp after applying the suggestions I posted above:

<details>
<summary>Preview <code>InsecureClassLoading.qhelp</code></summary>

# Android insecure context creation
Android Contexts that depend on external inputs are potentially vulnerable to code execution attacks. This query detects the scenarios where an attacker controlled input is used to create a new Android ` Context`.


## Recommendation
This can be avoided in two ways.

1. To avoid arbitrary code execution, the application should sanitize all external inputs.
1. Changing the application architecture to prevent unintended access to sensitive components.

## Example
The following example shows a 'bad' way of creating Android \`Context\` variables. An attacker may register a package with the name \`somename.plugin.\` having a \`pluginLoader\` class.


```java

import android.app.Application;
import android.content.Context;
import android.content.pm.PackageInfo;
import android.os.Bundle;

public class InsecureClassLoading extends Application {
  @Override
  public void onCreate() {
    super.onCreate();
    invokePlugins();
  }

  private void invokePlugins() {
    for (PackageInfo info : getPackageManager().getInstalledPackages(0)) {
      String packageName = info.packageName;
      Bundle meta = info.applicationInfo.metaData;
      if (packageName.startsWith("somename.plugin.")) {
        try {
          Context packageContext =
              createPackageContext(packageName, CONTEXT_INCLUDE_CODE | CONTEXT_IGNORE_SECURITY);
          packageContext.getClassLoader().loadClass("somename.pluginLoader")
              .getMethod("load", Context.class).invoke(null, this);
        } catch (Exception e) {
          throw new RuntimeException(e);
        }
      }
    }
  }
}

```
Then, when the code above executes, the attacker would be able to run malicious code in the context of the applciaiton.


## References
* Common Weakness Enumeration: [CWE-691](https://cwe.mitre.org/data/definitions/691.html).

</details>

porcupineyhairs replied:
The query rendering fails for me with a null pointer exception. Can you please take a look?
```
Oops! A fatal internal error occurred.
java.lang.NullPointerException
        at com.semmle.prettyprint.doc.QhelpToMD.checkFile(QhelpToMD.java:692)
        at com.semmle.prettyprint.doc.QhelpToMD.findFile(QhelpToMD.java:675)
        at com.semmle.prettyprint.doc.QhelpToMD.access$200(QhelpToMD.java:83)
        at com.semmle.prettyprint.doc.QhelpToMD$QhelpSaxHandler.startElement(QhelpToMD.java:389)
        at org.apache.xerces.parsers.AbstractSAXParser.startElement(Unknown Source)
        at org.apache.xerces.parsers.AbstractXMLDocumentParser.emptyElement(Unknown Source)
        at org.apache.xerces.impl.dtd.XMLDTDValidator.emptyElement(Unknown Source)
        at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanStartElement(Unknown Source)
        at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl$FragmentContentDispatcher.dispatch(Unknown Source)
        at org.apache.xerces.impl.XMLDocumentFragmentScannerImpl.scanDocument(Unknown Source)
        at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
        at org.apache.xerces.parsers.XML11Configuration.parse(Unknown Source)
        at org.apache.xerces.parsers.XMLParser.parse(Unknown Source)
        at org.apache.xerces.parsers.AbstractSAXParser.parse(Unknown Source)
        at org.apache.xerces.jaxp.SAXParserImpl$JAXPSAXParser.parse(Unknown Source)
        at org.apache.xerces.jaxp.SAXParserImpl.parse(Unknown Source)
        at com.semmle.prettyprint.doc.QhelpToMD.runSaxParse(QhelpToMD.java:814)
        at com.semmle.prettyprint.doc.QhelpToMD.process(QhelpToMD.java:764)
        at com.semmle.prettyprint.doc.QhelpToMD.format(QhelpToMD.java:627)
        at com.semmle.cli2.generate.QueryHelpCommand.executeSubcommand(QueryHelpCommand.java:134)
        at com.semmle.cli2.picocli.SubcommandCommon.call(SubcommandCommon.java:430)
        at com.semmle.cli2.picocli.SubcommandMaker.runMain(SubcommandMaker.java:201)
        at com.semmle.cli2.picocli.SubcommandMaker.runMain(SubcommandMaker.java:209)
        at com.semmle.cli2.CodeQL.main(CodeQL.java:91)
```

aibaars replied:
There is a bug in file path handling. A call to `Path.getParent` returns `null` for a path like `InsecureClassLoading.qhelp` causing a NullPointerException in `.QhelpToMD.checkFile`. Could you try calling the command with `./InsecureClassLoading.qhelp` as a workaround?

I'll look into fixing the bug properly. Thanks for reporting.

smowton commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.java:
```diff
@@ -0,0 +1,30 @@
+  }
+
+  private void invokePlugins() {
+    for (PackageInfo info : getPackageManager().getInstalledPackages(0)) {
```
The badness here is somewhat subtle, so suggest adding a comment -- the badness is that getInstalledPackages(...) enumerates packages belonging to other applications, so they could have dropped their own package that appears to be one of our plugins, right?

porcupineyhairs replied:
yes, I have added a comment in the code to reflect the same.

aibaars commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qhelp:
```diff
@@ -0,0 +1,33 @@
+  </overview>
+
+  <recommendation>
+    This can be avoided in two ways. 
```
```suggestion
    <p>This can be avoided in two ways. </p>
```

smowton commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.java:
```diff
@@ -0,0 +1,30 @@
+      if (packageName.startsWith("somename.plugin.")) {
+        try {
+          Context packageContext =
+              createPackageContext(packageName, CONTEXT_INCLUDE_CODE | CONTEXT_IGNORE_SECURITY);
```
Is the IGNORE_SECURITY flag here fundamental or merely incidental to the root problem?

smowton commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qhelp:
```diff
@@ -0,0 +1,33 @@
+
+  <overview>
+    <p>
+      Android Contexts that depend on external inputs are potentially vulnerable to code execution attacks.          
+      This query detects the scenarios where an attacker controlled input is used to create a new Android <code> Context</code>. 
```
```suggestion
      Loading a class that could have been installed by another Android application will lead to arbitrary code being executed in the context of the victim application.
      This query detects use of attacker-controlled input to create a new Android <code>Context</code>, which could be used to instantiate untrusted classes. 
```

aibaars replied:
```suggestion
      This query detects the scenarios where an attacker controlled input is used to create a new Android <code>Context</code>. 
```

smowton commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qhelp:
```diff
@@ -0,0 +1,33 @@
+    This can be avoided in two ways. 
+    <ol>
+      <li> 
+        To avoid arbitrary code execution, the application should sanitize all external inputs.
```
Be more specific -- what does sanitization look like in this case?

smowton commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qhelp:
```diff
@@ -0,0 +1,33 @@
+      <li> 
+        To avoid arbitrary code execution, the application should sanitize all external inputs.
+      </li>
+      <li> Changing the application architecture to prevent unintended access to sensitive components.</li>
```
Again be more specific -- what does a good architecture look like here?

porcupineyhairs replied:
This is difficult to explain in a `qhelp`. I have included a good example now though. PTAL.

smowton commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qhelp:
```diff
@@ -0,0 +1,33 @@
+  </recommendation>
+
+  <example>
+    <p>The following example shows a 'bad' way of creating Android `Context` variables. An attacker may register a package with the name `somename.plugin.` having a `pluginLoader`  class.</p>
```
```suggestion
    <p>The following example shows a dangerous way of creating an Android `Context`. A malicious, attacker-controlled app may have registered a package with the name `somename.plugin.xyz` having a `pluginLoader`  class, which the victim app will encounter and execute.</p>
```

aibaars replied:
```suggestion
    <p>The following example shows a 'bad' way of creating Android <code>Context</code> variables. An attacker may register a package with the name <code>somename.plugin.</code> having a <code>pluginLoader</code>  class.</p>
```

smowton commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qhelp:
```diff
@@ -0,0 +1,33 @@
+  <example>
+    <p>The following example shows a 'bad' way of creating Android `Context` variables. An attacker may register a package with the name `somename.plugin.` having a `pluginLoader`  class.</p>
+    <sample src="InsecureClassLoading.java" />
+    <p> Then, when the code above executes, the attacker would be able to run malicious code in the context of the applciaiton.</p>
```
Provide a good / fixed example too.

smowton commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qhelp:
```diff
@@ -0,0 +1,33 @@
+
+  <references>
+    <li>
+      <a href="https://cwe.mitre.org/data/definitions/749.html">CWE-749</a>
```
Is there any Google security advice about plugin architectures you can link here?

porcupineyhairs replied:
Unfortunately, I can find anything to suit my needs here so I am just ignoring it for now.

smowton commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoadingGood.java:
```diff
@@ -0,0 +1,55 @@
+
+        try {
+          CertificateFactory certFactory = CertificateFactory.getInstance("X509");
+          X509Certificate x509Cert = (X509Certificate) certFactory.generateCertificate(certStream);
```
To check: does this just get the presented certificate, or does it verify the chain of trust as well?

porcupineyhairs replied:
Only the presented certificate I think.

smowton reviewed (commented):
I like the good example, but I think this query wouldn't notice the cert check and would still flag that case, right? Either way you should add the good example to the test suite as well.

aschackmull commented on java/ql/src/semmle/code/java/dataflow/FlowSources.qll:
```diff
@@ -341,3 +341,17 @@ class ExportedAndroidIntentInput extends RemoteFlowSource, AndroidIntentInput {
+
+/** Android `PackageInfo` objects that may have come from a hostile application. */
+class AndroidPackageInput extends DataFlow::Node {
+  Type receiverType;
+
```
Unused field. Remove.
```suggestion
```

porcupineyhairs replied:
done!

smowton commented:
@porcupineyhairs I think you're still applying review comments? Please say when this is ready to re-review.

smowton commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qhelp:
```diff
@@ -0,0 +1,41 @@
+
+  <example>
+    <p>
+      The following example shows a dangerous way of creating an Android
+      <code>Context</code>
+      . A malicious, attacker-controlled app may have registered a package with the name
```
```suggestion
      The following example shows a dangerous way of creating an Android <code>Context</code>. 
      A malicious, attacker-controlled app may have registered a package with the name
```

smowton commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.qhelp:
```diff
@@ -0,0 +1,41 @@
+      The following example shows a dangerous way of creating an Android
+      <code>Context</code>
+      . A malicious, attacker-controlled app may have registered a package with the name
+      <code>somename.plugin.xyz</code>
+      having a
+      <code>pluginLoader</code>
+      class, which the victim app will encounter and execute.
```
```suggestion
      <code>somename.plugin.xyz</code> having a <code>pluginLoader</code> class, which the victim app will encounter and execute.
```

smowton reviewed (commented):
Took a quick look in case this is ready, not a lot more to say

porcupineyhairs commented:
@smowton I haven't made the changes you asked yet. I have broadened the scope of the query. The qhelp is yet to be updated to reflect the same. Since, I have concerns regarding the TP/FP ratio, can you run the query against LGTM again and share the results? Qhelp can be updated later.

intrigus-lgtm commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.ql:
```diff
@@ -30,13 +31,34 @@ class InsecureContextCreationConfig extends TaintTracking::Configuration {
   }
 }
 
+/** A sink for insecure context creation vulnerabilities */
```
This qldoc seems to be a duplicate of the `InsecureContextCreationSink` qldoc.

intrigus-lgtm commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.ql:
```diff
@@ -0,0 +1,119 @@
+/**  Models the constant `CONTEXT_INCLUDE_CODE` declared in `android.content.Context`class. */
+private class ContextIncludeCode extends FieldAccess {
+  ContextIncludeCode() {
+    exists(TypeContext t |
+      this.getField().getDeclaringType().extendsOrImplements*(t) and
+      this.getField().hasName("CONTEXT_INCLUDE_CODE")
+    )
```
The field is defined in `TypeContext` so we don't have to look in subclasses I'd say.

```suggestion
    this.getField().getDeclaringType() instanceof TypeContext and
    this.getField().hasName("CONTEXT_INCLUDE_CODE")
```

porcupineyhairs replied:
@intrigus-lgtm extending `Context` is not unheard of. There are multiple classes even in the Android OS code base doing so. I think keeping this as is is alright. This should have negligible performance impact anyways.

intrigus-lgtm replied:
But the constant is defined as a `public static final` constant in `android.content.Context`.
So any subclass would also have to use the constants from `android.content.Context`.
I don't think it's possible to pass constants - that are defined in subclasses of Context - to the `createPackageContext` method, or is it?
It would also be strange to redefine constants of a parent class.

intrigus-lgtm commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.ql:
```diff
@@ -0,0 +1,119 @@
+  }
+}
+
+/**  Models the constant `CONTEXT_IGNORE_SECURITY` declared in `android.content.Context`. */
```
```suggestion
/** Models an access to the constant `CONTEXT_IGNORE_SECURITY` declared in the `android.content.Context` class. */
```

intrigus-lgtm commented on java/ql/src/experimental/Security/CWE/CWE-691/InsecureClassLoading.ql:
```diff
@@ -0,0 +1,119 @@
+/**  Models the constant `CONTEXT_IGNORE_SECURITY` declared in `android.content.Context`. */
+private class ContextIgnoreSecurity extends FieldAccess {
+  ContextIgnoreSecurity() {
+    exists(TypeContext t |
+      this.getField().getDeclaringType().extendsOrImplements*(t) and
+      this.getField().hasName("CONTEXT_IGNORE_SECURITY")
+    )
```
The field is defined in `TypeContext` so we don't have to look in subclasses I'd say.
```suggestion
    this.getField().getDeclaringType() instanceof TypeContext and
    this.getField().hasName("CONTEXT_IGNORE_SECURITY")
    )
```

porcupineyhairs replied:
see my comment above. https://github.com/github/codeql/pull/4690#discussion_r570286403

porcupineyhairs commented:
@intrigus-lgtm @smowton @aschackmull @aibaars 
Please the discussion [here](https://github.com/github/securitylab/issues/230#issuecomment-778264596). The query has a very low security impact. Hence, I have closed the bounty application and am now closing this PR. 
My apologies for your time lost reviewing this query.
